docs/contributor/adr/0010-geoblocking-draft.md

```diff
@@ -0,0 +1,387 @@
+- performing configuration checks (like external traffic policy)
+- reporting geoblocking state
+
+There should be only one Geoblocking resource. The controller should raise an issue if there are more of them similarly to [APIGateway CR](https://github.com/kyma-project/api-gateway/blob/05ed57f3299f42c2565ed1b28b84dee808a5213b/controllers/operator/apigateway_controller.go#L100).
+
+![Geoblocking controller](../../assets/geoblocking-cr-controller.svg)
+
+#### CR examples
+
+##### Usage of SAP internal service
+
+```yaml
+apiVersion: geoblocking.kyma-project.io/v1alpha1
+kind: GeoBlocking
+metadata:
+  name: geoblocking
+  namespace: kyma-system
+spec:
+  gbService:
+    secret: "gb-secret"
+    tokenUrl: https://oauth.example.com/oauth2/token
+    ipRange:
+      url: https://lists.example.com/lists/some-list
+      refreshInSeconds: 3600
+    events:
+      url: https://events.example.com/events
+      lobId: "kyma"
+      productId: "some-product"
+      systemId: "some-system"
+```
+
+##### IP range allow/block list populated by customer
+```yaml
+apiVersion: geoblocking.kyma-project.io/v1alpha1
+kind: GeoBlocking
+metadata:
+    name: geoblocking
+    namespace: kyma-system
+spec:
+  ipRange:
+    configMap: "my-own-ip-range-list"
+```
+
+### Technical details
+
+#### Policy download optimization
+
+In order to reduce unnecessary updates of the list of blocked IP ranges the ip-auth service should use ETag mechanism, which is supported by the SAP internal service.
+
+The local copy (stored in a Config Map) should be used to optimize the amount of connections to the SAP internal service:
+- if the local copy contains a newer version (lastUpdateTime) -> load it
+- if the update check has been recently performed (now - lastUpdateCheckTime < refreshInterval) -> don't check it again
+- if the update check hasn't been recently performed (now - lastUpdateCheckTime >= refreshInterval) -> check for an update
+- if there is a newer version -> download, apply it, update a local copy and set lastUpdateCheckTime and lastUpdateTime to the current time
+- if there is no newer version -> update lastUpdateCheckTime time (so other pods may skip the check)
+- if there is a problem with checking for an update -> log a warning
+- if there is a problem with updating the local copy -> log a warning and use a new version anyway
+
+Pods should slightly randomize an update check time to benefit from the above optimization (minimize a risk of performing the check by multiple pods at the same time).
+
+#### Quick IP check
+
+The list of blocked IP ranges may be big (thousands of entries), so analyzing whether an IP address matches any IP range in the list is too time consuming.
+
+The list of IP ranges changes rarely, so it is better to build a data structure that supports quick comparision of IP addresses. The good candidate is a [radix tree](https://en.wikipedia.org/wiki/Radix_tree).
```

Comment:
The problem is that hash table is not applicable here, because the list of policies contains IP **ranges** and we need to check whether a given single IP address is inside any of ranges. So we would need a hashtable with 2^32 IP addresses (for IPv4)...
And the functionality we need seems to be available OOTB here: https://pkg.go.dev/github.com/infobloxopen/go-trees/iptree#Tree.GetByIP

Reply:
To underline - the ADR doesn't force any implementation here. It only suggests that the list of IP ranges may be huge and a linear search is not a good idea, because it has a complexity O(n). The radix tree was used as an example of something better than a 'raw' unsorted list.

Unfortunately, IP ranges may overlap, the IP range list is provided either by the user or by the SAP service and we don't control the data in any way. In particular, 'allowed' and 'blocked' ones may overlap as well (allowed take precedence). But AFAIR, it doesn't matter which particular rule was used to allow or block the traffic.

Reply:
Okay, I adapted the description to underline that radix-tree is only a proposal, mentioned also about TSS and wrote that this may be decided during implementation phase:
```Because the list of IP ranges changes rarely (compared to the frequency of incoming connections) it is recommended to build some efficient data structure that supports quick comparision of IP addresses. The good candidate is a [radix tree](https://en.wikipedia.org/wiki/Radix_tree) or something like [TSS](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=3aa1dd14e3d1c20d1f09b0ce0d4b4dd7b1190885). This can be decided during the implementation phase and would probably require some tests (benchmark, etc.).```